### PR TITLE
StampingVisitor will visit the callbacks and errbacks of the signature

### PR DESCRIPTION
**Pull request by Nusnus**
When a signature is stamped, this will allow the visitor to also stamp the signature's link and link_error callbacks automatically if they exists before the stamping

### CODE REVIEW

File: celery/canvas.py [hunk 1]
```diff
@@ -502,18 +525,48 @@ def stamp(self, visitor=None, **headers):
         headers = headers.copy()
         if visitor is not None:
             visitor_headers = visitor.on_signature(self, **headers)
-            if "stamped_headers" not in visitor_headers:
+            if visitor_headers and "stamped_headers" not in visitor_headers:
                 visitor_headers["stamped_headers"] = list(visitor_headers.keys())
-            headers.update(visitor_headers)
+            _merge_dictionaries(headers, visitor_headers or {})
```
Comment (thedrow) [inline celery/canvas.py:530]: When will `visitor_headers` be a falsy value?
528 checks if it isn't so why `or {}` is necessary?

File: t/unit/tasks/test_canvas.py [hunk 2]
```diff
@@ -636,10 +636,12 @@ def s(*args, **kwargs):
         assert c.tasks[-1].options['chord'] == 'some_chord_id'
 
         c.apply_async(link=[s(32)])
+        c.tasks[-1].options['link'][0].options.clear()  # remove stamped headers from link
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:639]: Why do we need this?

**Review by thedrow — CHANGES_REQUESTED** [review]
This needs more tests and documentation.
There are only minor issues I'd like to resolve.

File: celery/canvas.py [hunk 3]
```diff
@@ -502,18 +525,50 @@ def stamp(self, visitor=None, **headers):
...
+    def stamp_links(self, visitor, **headers):
+        """Stamp this signature links (callbacks and errbacks).
+        Using a visitor will pass on responsibility for the stamping
+        to the visitor.
+
+        Arguments:
+            visitor (StampingVisitor): Visitor API object.
+            headers (Dict): Stamps that should be added to headers.
+        """
+        additional_headers = headers.copy()
```
Comment (thedrow) [inline celery/canvas.py:550]: Can we find a more meaningful name?

Comment (Nusnus) [reply celery/canvas.py]: Fixed

File: t/unit/tasks/test_canvas.py (hunk 2 above)
Comment (Nusnus) [reply t/unit/tasks/test_canvas.py]: Fix Pushed @thedrow

**Comment by Nusnus** [comment]
Added doc to feature

**Review by thedrow — APPROVED** [review]
LGTM

**Pull request merged**